### ISSUE TITLE
How to find defended and attacked pieces?

### ISSUE THREAD

**Issue by vitogit** [jhlywa/chess.js#186]
What´s the best way to get the number of defenders and attackers for each square? 
I want to show in the board what pieces are unguarded and the "influence" for each square, this is the number of attackers - defenders for that square.
Maybe if the legal moves method had an option to include the squares that already have an own piece it could work.

**Comment by mliebelt** [comment]
I would like to have that functionality as well, I will give it a try. I see the following options (with different possible implementations then):

* 2 different calls, one for defended pieces (or squares), one for attacked pieces (or squares).
* each time the option to do that for the turn color, the other color, or both. Here an example
  * Searching for defended pieces, you have to simulate that first an attack is done, and then the defenders are computed. Currently, chess.js does not support defending moves.
  * Searching for attacking pieces, this will normally work in the current position for the color that has the turn. To get attacking pieces for the other color, you have to change the turn of the current position.

So the API could look like this:

* `attacked_pieces(color)`
  * color: default is null ==> denotes the one that has the turn.
  * BLACK | WHITE: defines for which color to search for attacked_pieces
  * BOTH: search for both colors
* `defended_pieces(color)` with the same semantics
* Both return something like `{'bd8': ['ne6', 'rd1'], 'qc7': ['ne6', 'bf4'], 'nf7': []}` ==> Bishop on d8 is attacked by Knight on e6 and by Rook on d1 ...

**Comment by vitogit** [comment]
I did it in a hacky way and used it here https://jsfiddle.net/m5q6fgtb/859/ (When you move a piece it show number of defenders and attackers per square).

To know the numbers of defenders for my piece (Same method can be used to know the attackers), I will remove that piece and put the enemy queen
and then check how many of my pieces can capture it, so that's the  number of defenders. Is not tested much so maybe it's wrong,but it can help someone.

```
function countSquareDefenders(boardElement, fen, square, me) {   
    var chess = new Chess(fenForOtherSide(fen));

    var oppositeColor = chess.turn() == 'w' ? 'b' : 'w'
    var queenSquare = squaresOfPiece(fen, oppositeColor, 'q');
    var somePiece = chess.remove(square); // remove my piece if any
    chess.remove(queenSquare); // remove his queen
    chess.put({ // put queen in the square
        type: 'q',
        color: oppositeColor
    }, square);    
  
    var moves = chess.moves({verbose: true, legal: false})
    var defendersCount = moves.filter(m => m.to == square && m.flags =='c').length
    
    if (somePiece && somePiece.color != oppositeColor) // count our own piece as controlling that square
    	defendersCount = parseInt(defendersCount)+1
      
    return defendersCount
}
```

**Comment by mliebelt** [comment]
I tried a different approach. I checked carefully the (internal) function `generate_moves`. All the logic is there, but by allowing `defending moves`, you get the moves that defend your pieces (or squares if you want). Not sure if that is the right way, but the unit tests are promising. I will hopefully commit my solution tomorrow, so you can check the pull request, if that is easier to use.

**Comment by captncraig** [comment]
I'd also love to have this information. I'm not sure using `generate_moves` is necessarily the right approach though? It seems if there are checks involved, there could be very few valid moves, but still a rather complex "influence" map. 

Should attacking and defending be based on a more naive move generation algorithm that doesn't account for check? In other words, look at *all* potential moves for all pieces, rather than just *valid* moves? Does that make any sense?

**Comment by hounaine** [comment]
I have done such work many years ago... manually.:) Sadly I do not code or only easy stuff. 
See if this approach can help. 
Each piece controls a certain number of cells. Those cells can be empty or occupied. When they are occupied by enemies it is called attacks. When they are occupied by our pieces it is called defence. Controlled empty cells+controlled occupied own pieces give the global defence of a player. 
This calculation should be done regularly and it serves many purposes. Ideally they must be a balance between attack and defence. There is other key concepts linked to the control concept. 
The questions we should agree about see numerous: is a cell controlled is attacked or defended? Should we take in account empty cells? When we talk about defence are we talking only our side? 
We need also to add the cumulate control (attack or defence): number of pieces that defend a cell or a piece?
We need to introduce the "indirect control": example: a queen in a same line of a rook. The queen controls indirectly the cells controlled by the rook. (This can never happen to knight and queen). 
Many things to say....

**Comment by DarthFloopy** [comment]
Has there been any progress on this since \**checks date*\* February? I am working on a chess game project and it would be awesome if there were a drop-in library that could tell what pieces are being attacked in a position. (I don't know much about this project. Can I do this with the current state of this project?)

**Comment by chao-mu** [comment]
I need this so bad. I made a game that is based around select undefended pieces and I have isAttacked to see if a piece is defended, but I need to know what it's defended by because a king can't defend if the opponent threatens that square.

**Comment by chao-mu** [comment]
```typescript
function getUndefendedChessmen(squareNames: Array<string>, fen: string) {
  const chess = new Chess(fen)
  const undefended = new Set<string>()

  const sightsByDest: Record<string, Set<Sight>> = {}
  for (const square of ALL_SQUARES) {
    sightsByDest[square] = new Set<Sight>()
  }

  for (const square of squareNames) {
    const piece = chess.get(square)
    if (!piece) {
      continue
    }

    chess.moves({ square: square, verbose: true }).forEach(move => {
      const sight = new Sight({
        piece: move.piece,
        color: move.color,
        square: move.to,
        origin: square
      })

      sightsByDest[move.to].add(sight)
    })
  }

  // Find undefended chessmen
  for (const square of squareNames) {
    const piece = chess.get(square)
    if (!piece || piece.type == "k") {
      continue
    }

    const sighters = Array.from(sightsByDest[square])
    const defenders = sighters.filter(sight => sight.color == piece.color)
    if (defenders.length == 0) {
      undefended.add(square)
      continue
    }

    // Can only be defended by a king if the square is not attacked by an enemy
    if (defenders.length == 1) {
      const defender = defenders[0]
      if (defender.piece == "k" && !sighters.some(sight => sight.color != piece.color)) {
        undefended.add(square)
        continue
      }
    }
  }

  return undefended
}
```

**Comment by jhlywa** [comment]
@chao-mu  I could supply something generic returning a list of attacking square. It's up to the user to filter out which pieces they don't want in the list (in your case the king). So something like:

```js
chess.load('r1bqkbnr/ppp2ppp/2np4/1B2p3/4P3/5N2/PPPP1PPP/RNBQK2R w KQkq - 0 4')

chess.attackers('d2', WHITE)
// -> ['c1', 'd1', 'e1', 'f3']

chess.attackers('d2', BLACK)
// -> []
```

How about the pinned knight on c6, does that get included?

```js
chess.attackers('d4', BLACK)
// -> ['c6', 'e5']   <-- should the c6 knight be included?
```

How about en-passant?

**Comment by chao-mu** [comment]
Whoops, deleted my post before seeing the response. There was a ton of bugs in it.

List of attacking squares would be perfect. I think attackers are only ones that can legally attack, so absolutely pinned pieces can't attack.

**Comment by chao-mu** [comment]
This is the direction I'm going for something like that (not working yet).

```typescript
function getAttackers(chess, square: string, color: string) {
  const attackers = Array.from(ALL_SQUARES).filter((attackerSquare) => {
    const attacker = chess.get(attackerSquare)
    if (!attacker || attacker.color != color) {
      return false
    }

    const moves = chess.moves({ square: attackerSquare, verbose: true })                                                                                                 
    return moves.some((move) => move.to == square)                                                                                                                       
  })                                                                                                                                                                     
                                         
  return new Set<string>(attackers)                                                                                                                                      
}                                                                                                                                                                        
```

**Comment by DarthFloopy** [comment]
What I ended up doing was just checking the list of legal moves for captures and using those destination squares.

As far as additions to the API, perhaps there could be separate methods to get a list of legal captures and a list of pieces “aiming” at a square (even if pinned). Maybe it would make sense to also have a list of squares a pieces is aiming at

**Comment by chao-mu** [comment]
I am not concerned with captures, just defenders. I'm expecting "attackers"  to return white attacking itself.

**Comment by chao-mu** [comment]
Essentially I want the contract for isAttacked, but returning the attackers

**Comment by neofight78** [comment]
At first sight it looks like #456 solves this. However, it doesn't take into account batteries. If a queen and bishop are in line pointing to one square, that will currently be counted as 1 attacker (or defender), but I presume people would want that to count as 2?

**Comment by axolotl-logic** [comment]
I would want it to count as 1 attacker. You couldn't capture with the piece on the other side of the battery.

**Comment by jhlywa** [comment]
I agree with @axolotl-logic. I think `attackers()` should count the number of immediate attackers.  We could always add an `{xray: true}` option (or something similar) to get the count for battery configurations.

**Comment by neofight78** [comment]
Fair enough. Don't need to change it if it's working how people expect / want.

**Comment by neofight78** [comment]
Looks like everything is covered then, if somebody needs something else, then they can raise a new issue with their specific needs.
